DOC: another loadmat example with a MATLAB struct

mikofski opened the pull request:
* show how MATLAB struct is loaded as [NumPy structured array](https://docs.scipy.org/doc/numpy/user/basics.rec.html#module-numpy.doc.structured_arrays) by default (not actually a [record array](https://docs.scipy.org/doc/numpy/user/basics.rec.html#module-numpy.doc.structured_arrays)) unless `struct_as_record=False`
* show how the size of the structured array is the same size as the MATLAB struct, and that its shape defaults to 2-D unless `squeeze_me=True`
* show how to index into the MATLAB struct to get fields and elements

@matthew-brett is this okay? is it too long? I cribbed liberally from my blog post that I linked to in [this comment](../../../scipy/scipy/pull/8762#issuecomment-383780860)

mikofski committed:
DOC: loadmat(<MATLAB-struct>) example

matthew-brett commented on scipy/io/matlab/mio.py:
```diff
@@ -157,6 +157,47 @@ def loadmat(file_name, mdict=None, appendmat=True, **kwargs):
             3.92699082, 4.71238898, 5.49778714, 6.28318531]])
+
+    By default SciPy reads MATLAB structures as structured NumPy arrays of
+    dtype object. This can be disabled by setting the optional arguement
```
"dtype object" is a bit obscure.  How about "structured NumPy arrays, where the fields in the structures become fields in the structured NumPy array."

mikofski replied:
OK, I changed the wording to:
>By default SciPy reads MATLAB structs as structured NumPy arrays where the
>dtypes are objects named after the MATLAB struct fields. This can be
>disabled by setting the optional argument `struct_as_record=False`.

IMO I think it's important to tell the user that the NumPy structured array fields will be *objects* which was not intuitive to me, I thought that a MATLAB struct field full of double arrays would be a equivalent to a named field in NumPy also of double arrays, but no.

Also I guess this raises the issue, that if *dtype object* is too obscure here, how will the user react when they see the MATLAB struct of strings, doubles and complex types become *objects* in NumPy? Perhaps there should be a cross reference to NumPy structured arrays or dtypes in see also? or elsewhere?

matthew-brett commented on scipy/io/matlab/mio.py:
```diff
@@ -157,6 +157,47 @@ def loadmat(file_name, mdict=None, appendmat=True, **kwargs):
+    )
+
+    The size of the structured array is the size of the MATLAB struct, not the
+    number elements in any particular field. The shape defaults to 2-D unless
```
number of elements in any

mikofski replied:
✔️

matthew-brett commented on scipy/io/matlab/mio.py:
```diff
@@ -157,6 +157,47 @@ def loadmat(file_name, mdict=None, appendmat=True, **kwargs):
+
+    The size of the structured array is the size of the MATLAB struct, not the
+    number elements in any particular field. The shape defaults to 2-D unless
+    the optional argument `squeeze_me=True`.
```
... in which case processing removes all length 1 dimensions (by the application of `np.squeeze`).

mikofski replied:
OK, I added a similar clarification, but I left out `np.squeeze` even tho it does actually use this in `mio_utils.pyx` it might not in the future? I just felt this was more info than was actually necessary. IMO the user needs to know that length 1 dimensions are dropped from the shape, but not necessarily how? Instead, I explained this further with another added example that shows what happens if you pass the arg `squeeze_me=True`.

matthew-brett reviewed (commented):
Thanks a lot - some small suggestions.

mikofski committed:
DOC: fix the doctest failure

* don't use assignment to test return, b/c will be None
* don't return big ugly array, save space just return dtypes to show they
objects with named fields same as MATLAB struct
* fix type, missing "of"
* expand on what "squeeze_me=True" does, add another example
* change backticks to quotes for MATLAB field names

Signed-off-by: Mark Mikofski <bwana.marko@yahoo.com>

mikofski commented:
I think I have addressed your concerns and also fixed my doctest error (I had an assignment that outputs nothing). I also tidied up the MATLAB struct example and added a `squeeze_me` example. Let me know if you have anymore comments.

matthew-brett commented on scipy/io/matlab/mio.py:
```diff
@@ -157,6 +157,52 @@ def loadmat(file_name, mdict=None, appendmat=True, **kwargs):
             3.92699082, 4.71238898, 5.49778714, 6.28318531]])
+
+    By default SciPy reads MATLAB structs as structured NumPy arrays where the
+    dtypes are objects named after the MATLAB struct fields. This can be
```
I'm really sorry, I see what you mean now, but I think this would be better phrased as "where the dtype field names correspond to the MATLAB struct field names, and each field is of `object` type".

matthew-brett replied:
By the way, I think the reason for this is because there is no requirement for the field values to be of the same type across structs in a Matlab struct array:

```matlab
>> s = struct('foo', {'bar', 1});
>> s.foo
ans = bar
ans =  1
```

mikofski replied:
Yes I agree, I think that this *is* the reason that the `dtype` of the MATLAB fields is `object`. Thanks so much for your review, and for letting me piggy back onto your PR. I hope that I've managed to add some value and not detract too much from your original intent.

matthew-brett reviewed (commented):
Some small comments - but otherwise - thanks - looks nice.

mikofski committed:
DOC: improve wording re: struct fields type and name

matthew-brett commented:
Excellent - thanks - this is a big improvement on my original.
